hooks/request/types.go

```diff
@@ -0,0 +1,248 @@
+	Calldata            string               `gorm:"index:idx_min_count_ask_count_oracle_script_id_calldata;size:1024"`
+	RequestedValidators []RequestedValidator `gorm:"constraint:OnDelete:CASCADE"`
+	MinCount            uint64               `gorm:"index:idx_min_count_ask_count_oracle_script_id_calldata"`
+	AskCount            uint64               `gorm:"index:idx_min_count_ask_count_oracle_script_id_calldata"`
+	AnsCount            uint64
+	RequestHeight       int64
+	RequestTime         time.Time
+	ClientID            string
+	RawRequests         []RawRequest `gorm:"constraint:OnDelete:CASCADE"`
+	IBCPortID           string
+	IBCChannelID        string
+	ExecuteGas          uint64
+	Reports             []Report  `gorm:"constraint:OnDelete:CASCADE"`
+	ResolveTime         time.Time `gorm:"index:idx_min_count_ask_count_oracle_script_id_calldata"`
+	// ResolveStatus is proto's resolve_status but as enum string
+	ResolveStatus string `gorm:"size:64"`
+	// Result is proto's Result.result field encoded in OBI and base64
+	Result string `gorm:"size:1024"`
+}
+
+// QueryRequestResponse convert GORM's Request model to proto's QueryRequestResponse
+func (r Request) QueryRequestResponse() types.QueryRequestResponse {
+	// Request's calldata
+	calldata, err := base64.StdEncoding.DecodeString(r.Calldata)
+	if err != nil {
+		panic(err)
+	}
+
+	// Requested validators
+	var requestedValidators []string
+	for _, rVal := range r.RequestedValidators {
+		requestedValidators = append(requestedValidators, rVal.Address)
+	}
+
+	// Raw requests
+	var rawRequests []types.RawRequest
+	for _, rr := range r.RawRequests {
+		calldata, err := base64.StdEncoding.DecodeString(rr.Calldata)
+		if err != nil {
+			panic(err)
+		}
+		rawRequests = append(rawRequests, types.RawRequest{
+			ExternalID:   types.ExternalID(rr.ExternalID),
+			DataSourceID: types.DataSourceID(rr.DataSourceID),
+			Calldata:     calldata,
+		})
+	}
+
+	// Result data
+	requestResult, err := base64.StdEncoding.DecodeString(r.Result)
+	if err != nil {
+		panic(err)
+	}
+
+	// Reports
+	var reports []types.Report
+	for _, dbReport := range r.Reports {
+		// Raw reports
+		var rawReports []types.RawReport
+		for _, rr := range dbReport.RawReports {
+			data, err := base64.StdEncoding.DecodeString(rr.Data)
+			if err != nil {
+				panic(err)
+			}
+			rawReports = append(rawReports, types.RawReport{
+				ExternalID: types.ExternalID(rr.ExternalID),
+				Data:       data,
+				ExitCode:   rr.ExitCode,
+			})
+		}
+		report := types.Report{
+			Validator:       dbReport.Validator,
+			InBeforeResolve: dbReport.InBeforeResolve,
+			RawReports:      rawReports,
+		}
+		reports = append(reports, report)
+	}
+
+	// IBC Channel
+	var ibcChannel *types.IBCChannel
+	if len(r.IBCChannelID) > 0 {
+		ibcChannel = &types.IBCChannel{
+			PortId:    r.IBCPortID,
+			ChannelId: r.IBCChannelID,
+		}
+	}
+
+	// The whole response
+	request := types.QueryRequestResponse{
+		Request: &types.Request{
```

Comment:
I think we using NewRequest to prevent a bug from missing some fields

Reply:
I have added a condition to check whether request exists or not.

Reply:
Fixed. It's now using function to create new instances.